modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/prepare/PrepareServiceImpl.java

```diff
@@ -984,4 +1306,87 @@ boolean matches() {
             return matches;
         }
     }
+
+    static class PlanningContextState {
+        private final Int2ObjectMap<Object> parameters;
+
+        PlanningContextState(
+                Int2ObjectMap<Object> parameters
+        ) {
+            this.parameters = parameters;
+        }
+
+        Int2ObjectMap<Object> parameters() {
+            return parameters;
+        }
+
+        /** Get context builder. */
+        public static PlanningContextState.Builder builder() {
+            return new PlanningContextState.Builder();
+        }
+
+        public static class Builder {
+            private Int2ObjectMap<Object> parameters;
+
+            /** Values of dynamic parameters to assist with type inference. */
+            PlanningContextState.Builder parameters(Int2ObjectMap<Object> parameters) {
+                this.parameters = parameters;
+                return this;
+            }
+
+            /** Builds planner context state. */
+            PlanningContextState build() {
+                return new PlanningContextState(parameters);
+            }
+        }
+    }
+
+    static class PlanInfo {
+        private final QueryPlan queryPlan;
+        @Nullable private final ValidStatement<ValidationResult> statement;
+        @Nullable private final PlanningContextState context;
+        @Nullable private final SchemaPlus defaultSchema;
+        private final IntSet sources;
+        private volatile boolean needToInvalidate;
+
+        private PlanInfo(
+                QueryPlan plan,
+                @Nullable ValidStatement<ValidationResult> statement,
+                @Nullable PlanningContextState context,
+                @Nullable SchemaPlus defaultSchema,
```

Comment:
Can we get SchemaPlus from root schema by name (from CacheKey) each time we do re-planning, instead of caching SchemaPlus.